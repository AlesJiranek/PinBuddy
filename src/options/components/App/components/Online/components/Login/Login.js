import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { userLogInAction } from 'redux-options/actions/user';
import { Paragraph, Error, Input, Button } from 'theme';
import Loading from './components/Loading';

class Login extends Component {
  state = {
    token: '',
  };

  static validateInput(input) {
    const regex = /^([^:]{1,})(:)(\S{1,})/;
    return input.match(regex);
  }

  render() {
    return (
      <form onSubmit={this.onFormSubmit}>
        <Paragraph innerHTML t={chrome.i18n.getMessage('optionsLoginManual')} />

        { this.props.error && <Error t={chrome.i18n.getMessage('optionsLoginError')} /> }

        {
          this.props.loading ? <Loading /> : (
            <>
              <Input
                id="token"
                label={chrome.i18n.getMessage('optionsLoginFormLabel')}
                placeholder={chrome.i18n.getMessage('optionsLoginFormPlaceholder')}
                onChange={this.onChangeUserInput}
                autoComplete="off"
                autoCorrect="off"
                autoCapitalize="off"
                spellCheck="false"
                autoFocus
              />

              <Button
                t={chrome.i18n.getMessage('optionsLoginFormSubmit')}
                type="submit"
                disabled={!Login.validateInput(this.state.token)}
              />
            </>
          )
        }
      </form>
    );
  }

  onFormSubmit = e => {
    e.preventDefault();
    this.props.userLogIn(this.state.token);
  }

  onChangeUserInput = e => {
    const { value } = e.target;
    this.setState(() => {
      return {
        token: value,
      };
    });
  }
}

Login.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.bool.isRequired,
  userLogIn: PropTypes.func.isRequired,
};

const mapStateToProps = state => ({
  loading: state.user.loading,
  error: state.user.error,
});

const mapDispatchToProps = dispatch => ({
  userLogIn: token => dispatch(userLogInAction(token))
});

export default connect(mapStateToProps, mapDispatchToProps)(Login);
