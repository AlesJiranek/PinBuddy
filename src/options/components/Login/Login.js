import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import { authenticateUser } from '../../actions/auth';
import { Paragraph, Error, Input, Button } from 'theme';

const API = 'https://api.pinboard.in/v1/';

class Login extends Component {
  state = {
    token: '',
    error: false,
    loading: false,
  };

  static validateInput(input) {
    const regex = /^([^:]{1,})(:)(\S{1,})/;
    return input.match(regex);
  }

  render() {
    return (
      <form onSubmit={this.onFormSubmit}>
        <Paragraph innerHTML t={chrome.i18n.getMessage('optionsLoginManual')} />

        { this.state.error && <Error t={chrome.i18n.getMessage('optionsLoginError')} /> }

        {
          this.state.loading ? (
            <Paragraph innerHTML t={chrome.i18n.getMessage('optionsLoginLoading')} />
          ) : (
            <>
              <Input
                id="token"
                label={chrome.i18n.getMessage('optionsLoginFormLabel')}
                placeholder={chrome.i18n.getMessage('optionsLoginFormPlaceholder')}
                onChange={this.onChangeUserInput}
              />

              <Button
                t={chrome.i18n.getMessage('optionsLoginFormSubmit')}
                type="submit"
                disabled={!Login.validateInput(this.state.token)}
              />
            </>
          )
        }
      </form>
    );
  }

  onFormSubmit = e => {
    e.preventDefault();
    this.setState({
      loading: true,
      error: false,
    });

    fetch(`${API}user/api_token?format=json&auth_token=${this.state.token}`)
      .then(dataAuth => dataAuth.json())
      .then(() => {
        const now = Date.now();
        const [username, token] = this.state.token.split(':');
        chrome.storage.local.set(
          {
            username,
            token,
          },
          () => {
            this.props.authenticateUser(this.state.token);
          }
        );

        fetch(`${API}posts/all?format=json&auth_token=${this.state.token}`)
          .then(dataPosts => dataPosts.json())
          .then(posts => {
            chrome.storage.local.set(
              {
                posts,
                postsFetched: now,
              }
            );
          });

        fetch(`${API}tags/get?format=json&auth_token=${this.state.token}`)
          .then(dataTags => dataTags.json())
          .then(tags => {
            const listOfTags = Object.keys(tags);

            chrome.storage.local.set(
              {
                tags: listOfTags,
                tagsFetched: now,
              }
            );
          });

      })
      .catch(() => {
        this.setState({
          error: true,
          loading: false,
        });
      });
  }

  onChangeUserInput = e => {
    const { value } = e.target;
    this.setState(() => {
      return {
        token: value,
      };
    });
  }
}

Login.propTypes = {
  token: PropTypes.string.isRequired,
  username: PropTypes.string.isRequired,
  authenticateUser: PropTypes.func.isRequired,
};

const mapStateToProps = (state, ownProps) => {
  return {
    token: state.auth.token,
    username: state.auth.username,
  };
};

const mapDispatchToProps = dispatch => {
  return {
    authenticateUser: token => {
      dispatch(authenticateUser(token));
    }
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(Login);
