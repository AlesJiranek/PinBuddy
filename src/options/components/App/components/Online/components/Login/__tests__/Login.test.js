import React from 'react';
import { shallow } from 'enzyme';

import { Login } from '../Login';

describe('<Login />', () => {

  const defaultProps = {
    loading: false,
    error: false,
    userLogIn: jest.fn(),
  };

  it('should render correctly', () => {
    const tree = shallow(
      <Login
        {...defaultProps}
      />
    );
    expect(tree).toMatchSnapshot();
  });

});
