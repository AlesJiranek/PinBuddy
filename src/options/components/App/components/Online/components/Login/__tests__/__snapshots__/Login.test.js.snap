// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> should render correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Login
    error={false}
    loading={false}
    userLogIn={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <Paragraph
          innerHTML={true}
          t="optionsLoginManual"
        />,
        false,
        <React.Fragment>
          <ForwardRef
            autoCapitalize="off"
            autoComplete="off"
            autoCorrect="off"
            autoFocus={true}
            id="token"
            invalidMessage={null}
            label="optionsLoginFormLabel"
            onChange={[Function]}
            pattern={null}
            placeholder="optionsLoginFormPlaceholder"
            spellCheck="false"
            type="text"
          />
          <Button
            disabled={true}
            onClick={null}
            secondary={false}
            t="optionsLoginFormSubmit"
            title={null}
            type="submit"
          />
        </React.Fragment>,
      ],
      "onSubmit": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "innerHTML": true,
          "t": "optionsLoginManual",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      false,
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            <ForwardRef
              autoCapitalize="off"
              autoComplete="off"
              autoCorrect="off"
              autoFocus={true}
              id="token"
              invalidMessage={null}
              label="optionsLoginFormLabel"
              onChange={[Function]}
              pattern={null}
              placeholder="optionsLoginFormPlaceholder"
              spellCheck="false"
              type="text"
            />,
            <Button
              disabled={true}
              onClick={null}
              secondary={false}
              t="optionsLoginFormSubmit"
              title={null}
              type="submit"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "autoCapitalize": "off",
              "autoComplete": "off",
              "autoCorrect": "off",
              "autoFocus": true,
              "id": "token",
              "invalidMessage": null,
              "label": "optionsLoginFormLabel",
              "onChange": [Function],
              "pattern": null,
              "placeholder": "optionsLoginFormPlaceholder",
              "spellCheck": "false",
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": Object {
              "$$typeof": Symbol(react.forward_ref),
              "defaultProps": Object {
                "autoCapitalize": null,
                "autoComplete": null,
                "autoCorrect": null,
                "autoFocus": null,
                "invalidMessage": null,
                "pattern": null,
                "spellCheck": null,
                "type": "text",
              },
              "propTypes": Object {
                "autoCapitalize": [Function],
                "autoComplete": [Function],
                "autoCorrect": [Function],
                "autoFocus": [Function],
                "id": [Function],
                "invalidMessage": [Function],
                "label": [Function],
                "onChange": [Function],
                "pattern": [Function],
                "spellCheck": [Function],
                "type": [Function],
              },
              "render": [Function],
            },
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "disabled": true,
              "onClick": null,
              "secondary": false,
              "t": "optionsLoginFormSubmit",
              "title": null,
              "type": "submit",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": Symbol(react.fragment),
      },
    ],
    "type": "form",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <Paragraph
            innerHTML={true}
            t="optionsLoginManual"
          />,
          false,
          <React.Fragment>
            <ForwardRef
              autoCapitalize="off"
              autoComplete="off"
              autoCorrect="off"
              autoFocus={true}
              id="token"
              invalidMessage={null}
              label="optionsLoginFormLabel"
              onChange={[Function]}
              pattern={null}
              placeholder="optionsLoginFormPlaceholder"
              spellCheck="false"
              type="text"
            />
            <Button
              disabled={true}
              onClick={null}
              secondary={false}
              t="optionsLoginFormSubmit"
              title={null}
              type="submit"
            />
          </React.Fragment>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "innerHTML": true,
            "t": "optionsLoginManual",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        false,
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              <ForwardRef
                autoCapitalize="off"
                autoComplete="off"
                autoCorrect="off"
                autoFocus={true}
                id="token"
                invalidMessage={null}
                label="optionsLoginFormLabel"
                onChange={[Function]}
                pattern={null}
                placeholder="optionsLoginFormPlaceholder"
                spellCheck="false"
                type="text"
              />,
              <Button
                disabled={true}
                onClick={null}
                secondary={false}
                t="optionsLoginFormSubmit"
                title={null}
                type="submit"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "autoCapitalize": "off",
                "autoComplete": "off",
                "autoCorrect": "off",
                "autoFocus": true,
                "id": "token",
                "invalidMessage": null,
                "label": "optionsLoginFormLabel",
                "onChange": [Function],
                "pattern": null,
                "placeholder": "optionsLoginFormPlaceholder",
                "spellCheck": "false",
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": Object {
                "$$typeof": Symbol(react.forward_ref),
                "defaultProps": Object {
                  "autoCapitalize": null,
                  "autoComplete": null,
                  "autoCorrect": null,
                  "autoFocus": null,
                  "invalidMessage": null,
                  "pattern": null,
                  "spellCheck": null,
                  "type": "text",
                },
                "propTypes": Object {
                  "autoCapitalize": [Function],
                  "autoComplete": [Function],
                  "autoCorrect": [Function],
                  "autoFocus": [Function],
                  "id": [Function],
                  "invalidMessage": [Function],
                  "label": [Function],
                  "onChange": [Function],
                  "pattern": [Function],
                  "spellCheck": [Function],
                  "type": [Function],
                },
                "render": [Function],
              },
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "disabled": true,
                "onClick": null,
                "secondary": false,
                "t": "optionsLoginFormSubmit",
                "title": null,
                "type": "submit",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": Symbol(react.fragment),
        },
      ],
      "type": "form",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
